src/Microsoft.AspNet.Mvc.Razor/RazorViewEngine.cs

```diff
@@ -233,14 +234,30 @@ public virtual IEnumerable<string> AreaViewLocationFormats
             string pageName,
             bool isPartial)
         {
-            if (IsApplicationRelativePath(pageName))
+            var applicationRelativePath = pageName;
+            if (!IsApplicationRelativePath(applicationRelativePath) && IsRelativePath(applicationRelativePath))
             {
-                var applicationRelativePath = pageName;
-                if (!pageName.EndsWith(ViewExtension, StringComparison.OrdinalIgnoreCase))
+                // Given a relative path (i.e. ending with ".cshtml") that is not yet application-relative (i.e.
+                // starting with "~/" or "/"), interpret given path relative to currently-executing view, if any.
+                var executingPath = (context as ViewContext)?.ExecutingFilePath;
```

Comment:
This still feels very weird to me - it relies heavily on the way we construct `ViewContext` instances from `ActionContext` which happens to be an implementation detail and not a contract. How did this work for partials in the past? Did we resolve it in the HtmlHelper? 

cc @Eilon for thoughts?